### ISSUE TITLE
all_allowed_actions returns denied actions

### ISSUE THREAD

**Issue by fruechel** [salesforce/cloudsplaining#97]
Calling something like `RoleDetail.all_allowed_actions` will return denied statements, too.

https://github.com/salesforce/cloudsplaining/blob/master/cloudsplaining/scan/role_details.py#L153 -> https://github.com/salesforce/cloudsplaining/blob/master/cloudsplaining/scan/policy_document.py#L49 -> https://github.com/salesforce/cloudsplaining/blob/master/cloudsplaining/scan/statement_detail.py#L144

I think there are two possible places to fix this:

1. In PolicyDocument because its property name is very explicit
2. In StatementDetail because the `expanded_actions` docstring says "Expands the full list of allowed actions from the Policy/"

Because it's not obvious which of those is better, I didn't make a PR so let me know what's better and I'll make a fix.

FWIW the reason I stumbled on this was a policy that denies based on token timestamp:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Deny",
            "Action": [
                "*"
            ],
            "Resource": [
                "*"
            ],
            "Condition": {
                "DateLessThan": {
                    "aws:TokenIssueTime": "<time>"
                }
            }
        }
    ]
}```

**Comment by kmcquade** [comment]
Yep. So, one of my design decisions was to completely ignore conditions knowing that it would allow for false positives. I know that sounds crazy, but the amount of scrutiny that [Parliament](https://github.com/duo-labs/parliament/) went through to support conditions logic, and the maintenance that it requires, was not what I intended for the tool.

For example, the above policy would not increase anyone's permissions - it only places requirements on what must happen in order for the other allowed API calls to be permitted.

But I don't think you are asking for this logic to be parsed and have its complex voodoo logic understood by Cloudsplaining - I think you are just asking for it to *not return False Positives* - right?

In the interest of doing so, perhaps we can do the following:
* Skip the Statement ID if it is has `Deny` in it.
* In the authorization details results JSON output, we could add on another field that includes important logs that the user might want to consider in the report. 
  * We could have an entry that essentially says, "This Policy contains a statement with a Deny Effect, or it uses the NotAction or NotResource elements which can be risky depending on how they are leveraged." 
  * Then we have the report transform that data and render it inside a "scan log" tab. 
  * It can also contain a summary of where exclusions were applied, etc. 

https://github.com/salesforce/cloudsplaining/blob/cf040e20474564749b625f7c0642fdbaaaafae63/cloudsplaining/scan/authorization_details.py#L49-L61

What do you think? Does this approach make sense?

**Comment by fruechel** [comment]
@kmcquade thanks for the explanation. I'm happy to lean whichever way you want, however, I think that the API should reflect what it really does. For example, if we update the names and docstrings to reflect that it just expands actions (and doesn't regard conditions or effects), the user knows to expect this.

I'm actually using these APIs directly to produce a different kind of report, so I'm calling `role_detail.all_allowed_actions` from my own code, which is how I stumbled upon these. I tried to find where this kind of behaviour was introduced, but I couldn't find it. The old UI doesn't display this finding, but the new one does. I'm not quite sure why, but the policy in question is an inline policy and I think the old UI only does managed policies? Not sure.

As an aside, I also found that it might be useful to resolve from a Role to _which_ policy introduced risks. If there are complex policies and weird conditions, it might help the user determine which policy lead to the finding. Is this something you'd want? Should I create a separate issue for it?

On your proposed solution: I think skipping the statement if it has a Deny is a great and simple approach. In fact, it's what I did to fix it locally:

```
    @property
    def expanded_actions(self):
        """Expands the full list of allowed actions from the Policy/"""
        if self.effect_deny:
            return []
```

This is where I put my fix but you could easily put it one layer higher and then skip this call entirely. What do you think?

**Comment by kmcquade** [comment]
@fruechel - I believe this was fixed in #177, thanks to @njgibbon. Can you give it a shot and let me know if it works for you?

**Comment by fruechel** [comment]
I actually think that #166 fixed it already. I just created a quick test file and d695b096 definitely fixed the issue as it is still present in the previous commit. Regardless, I'm closing this issue since it's fixed now.
